File: frontend/packages/core/src/types/billing.ts
export interface CostAggregateRow {
  bucket: string
  bucket_type: 'workspace' | 'user' | 'model' | 'day'
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
  cost_amount_micro: number
  currency: string
  call_count: number
}

export interface CostSummaryResponse {
  from_date: string
  to_date: string
  total_cost_amount_micro: number
  currency: string
  total_calls: number
  by_workspace: CostAggregateRow[]
  by_model: CostAggregateRow[]
  by_user: CostAggregateRow[]
  by_day: CostAggregateRow[]
}

export interface TimeseriesPoint {
  date: string
  cost_amount_micro: number
  calls: number
  input_tokens: number
  output_tokens: number
  cache_read_tokens: number
  cache_write_tokens: number
}

export interface TimeseriesSeries {
  bucket: string
  points: TimeseriesPoint[]
  currency: string
}

export interface TimeseriesResponse {
  from_date: string
  to_date: string
  granularity: 'day' | 'week'
  dimension: 'workspace' | 'model' | 'user'
  series: TimeseriesSeries[]
  currency: string
}

export function formatCostUsd(micro: number, currency: string = 'USD'): string {
  const amount = micro / 1_000_000
  return `${currency} ${amount.toFixed(4)}`
}

File: backend/cubeplex/api/schemas/billing.py
"""Pydantic schemas for billing/cost API responses."""

from datetime import date
from typing import Literal

from pydantic import BaseModel


class CostAggregateRow(BaseModel):
    bucket: str  # workspace_id | user_id | "provider/model_id" | "YYYY-MM-DD"
    bucket_type: str  # "workspace" | "user" | "model" | "day"
    # Human-readable label for workspace/user buckets; None for model/day or
    # when the entity no longer exists (clients should fall back to `bucket`).
    bucket_label: str | None = None
    input_tokens: int
    output_tokens: int
    cache_read_tokens: int
    cache_write_tokens: int
    cost_amount_micro: int  # amount × 10⁶; divide by 1_000_000 for display
    currency: str
    call_count: int


class CostSummaryResponse(BaseModel):
    from_date: date
    to_date: date
    total_cost_amount_micro: int
    currency: str
    total_calls: int
    by_workspace: list[CostAggregateRow]
    by_model: list[CostAggregateRow]
    by_user: list[CostAggregateRow]
    by_day: list[CostAggregateRow]


class TimeseriesPoint(BaseModel):
    date: str  # YYYY-MM-DD (or week-start date if granularity=week)
    cost_amount_micro: int
    calls: int
    input_tokens: int
    output_tokens: int
    cache_read_tokens: int
    cache_write_tokens: int


class TimeseriesSeries(BaseModel):
    bucket: str  # workspace_id | user_id | "provider/model_id" | "__other"
    # Human-readable label for workspace/user series; None for model/__other or
    # when the entity no longer exists (clients should fall back to `bucket`).
    bucket_label: str | None = None
    points: list[TimeseriesPoint]
    currency: str


class TimeseriesResponse(BaseModel):
    from_date: date
    to_date: date
    granularity: Literal["day", "week"]
    dimension: Literal["workspace", "model", "user"]
    series: list[TimeseriesSeries]
    currency: str
